zapcore/sampler.go

```diff
@@ -62,28 +71,68 @@ func fnv32a(s string) uint32 {
 	return hash
 }
 
-func (c *counter) IncCheckReset(t time.Time, tick time.Duration) uint64 {
+func (c *counter) IncCheckReset(t time.Time, tick time.Duration) (currentVal, preResetVal uint64) {
 	tn := t.UnixNano()
 	resetAfter := c.resetAt.Load()
 	if resetAfter > tn {
-		return c.counter.Inc()
+		return c.counter.Inc(), 0
 	}
 
-	c.counter.Store(1)
+	preResetVal = c.counter.Load()
+	currentVal = 1
+	c.counter.Store(currentVal)
 
 	newResetAfter := tn + tick.Nanoseconds()
 	if !c.resetAt.CAS(resetAfter, newResetAfter) {
 		// We raced with another goroutine trying to reset, and it also reset
 		// the counter to 1, so we need to reincrement the counter.
-		return c.counter.Inc()
+		return c.counter.Inc(), 0
 	}
+	return currentVal, preResetVal
+
+}
+
+// Reporter defines report function that generates and writes sampling activity
+// report log entries.
+type Reporter interface {
```

Comment:
this almost certainly doesn't need to be exported (its only method is unexported, so it can't be implemented outside the zapcore package anyhow; which is the right thing to do, since we really don't want to grow zap's supported surface area if possible)

Reply:
The idea is to allow users write their own Reporter if/when there's a need to customize sampling report entry

Reply:
since we don't export this functionality today, let's unexport till needed